src/main/scala/com/yotpo/metorikku/test/Tester.scala

```diff
@@ -74,162 +74,407 @@ case class Tester(config: TesterConfig) {
-    val fieldResult = currFieldType match {
-      case x if x == TimestampType => StringType
-      case x if x == LongType => IntegerType
-      case x if x.toString.contains("Array") => StringType
-      case x if x.toString.contains("Object") => StringType
-      case x => x
+  private def logSubtablesErrors(sortedExpectedResults: List[Map[String, Any]], sortedActualResults: List[Map[String, Any]],
+                                 errorsIndexArrExpected: Seq[Int], errorsIndexArrActual: Seq[Int], redirectDfShowToLogger: Boolean): Unit = {
+    val isExpectedErrors = errorsIndexArrExpected.size > 1 //whitespace line
+    val isActualErrors = errorsIndexArrActual.size > 1
+    val expectedKeys = "row_number" +: sortedExpectedResults.head.keys.toList
+    if (isExpectedErrors) {
+      log.warn("**********************  Expected with Mismatches  ************************")
+      val subExpectedError = TestUtil.getSubTable(sortedExpectedResults, errorsIndexArrExpected.sorted)
+      showDfToConsoleOrLogger(redirectDfShowToLogger, transformListMapToDfWitIdCol(subExpectedError, expectedKeys),
+                                errorsIndexArrExpected.size, truncate = false, sortedExpectedResults.size)
+    }
+    if (isActualErrors) {
+      log.warn("***********************  Actual with Mismatches  *************************")
+      val subActualError = TestUtil.getSubTable(sortedActualResults, errorsIndexArrActual.sorted)
+      showDfToConsoleOrLogger(redirectDfShowToLogger = true, transformListMapToDfWitIdCol(subActualError, expectedKeys),
+                                errorsIndexArrActual.size, truncate = false, sortedActualResults.size)
     }
-    fieldResult
   }
 
-  private def matchExpectedRow(mapOfActualRow: Map[String, Nothing], metricExpectedResultRows: List[Map[String, Any]]): Map[String, Any] = {
-    // scalastyle:off
-    for (expectedRowCandidate <- metricExpectedResultRows) {
-      if (isMatchingValuesInRow(mapOfActualRow, expectedRowCandidate)) {
-        return expectedRowCandidate
-      }
+  private def transformListMapToDfWitIdCol(mapList: List[mutable.LinkedHashMap[String, Any]], schemaKeys: List[String]): DataFrame = {
+    val rowIdField = "row_number"
+    val mapSchemaKeysList = KeyColumns.removeUnexpectedColumns(mapList, schemaKeys)
+    val mapStrList = mapSchemaKeysList.map(x => x.mapValues(v => {
+      if (v == null) "" else v.toString
+    }
+    ))
+    val rows = mapStrList.map(m => spark.sql.Row(m.values.toSeq: _*))
+    val x: java.util.List[Row] = scala.collection.JavaConversions.seqAsJavaList(rows)
+    val schema = org.apache.spark.sql.types.StructType(schemaKeys.map(fieldName => StructField(fieldName, StringType, nullable = true)))
+    val df = job.sparkSession.createDataFrame(x, schema)
+    val isColContained = schemaKeys.contains(rowIdField)
+    val indexedDf = isColContained match {
+      case false => df.withColumn(rowIdField, monotonically_increasing_id() + 1)
```

Comment:
monotonically_increasing_id is not a good way to do 1...n .
You need to use a window function (row_number) .
monotonically_increasing_id can be very high numbers sometimes depends on the cluster.